message_ix/macro.py

```diff
@@ -1,61 +1,464 @@
+        if np.isclose(model_price['lvl'], 0).any():
+            # TODO: this needs a test..
+            msg = '0-price found in MESSAGE variable PRICE_COMMODITY'
+            raise RuntimeError(msg)
+        model_price.rename(columns={'lvl': 'value', 'commodity': 'sector'},
+                           inplace=True)
+        model_price = model_price[idx + ['value']]
+
+        # get data provided in init year from data
+        price_ref = self.data['price_ref'].reset_index()
+        price_ref['year'] = self.init_year
+        # combine into one value
+        price = pd.concat([price_ref, model_price]).set_index(idx)['value']
+        if price.isnull().any():
+            raise RuntimeError('NaN values found in price calculation')
+        self.data['price'] = price
+        return price
+
+    @lru_cache()
+    def _demand(self):
+        # read from scenario
+        idx = ['node', 'sector', 'year']
+        model_demand = self.s.var('DEMAND', filters={'level': 'useful'})
+        model_demand.rename(columns={'lvl': 'value', 'commodity': 'sector'},
+                            inplace=True)
+        model_demand = model_demand[idx + ['value']]
+        # get data provided in init year from data
+        demand_ref = self.data['demand_ref'].reset_index()
+        demand_ref['year'] = self.init_year
+        # combine into one value
+        demand = pd.concat([demand_ref, model_demand]).set_index(idx)['value']
+        if demand.isnull().any():
+            raise RuntimeError('NaN values found in demand calculation')
+        self.data['demand'] = demand
+        return demand
+
+    @lru_cache()
+    def _bconst(self):
+        price_ref = self.data['price_ref']
+        demand_ref = self.data['demand_ref']
+        rho = self._rho()
+        gdp0 = self._gdp0()
+        # TODO: automatically get the units here!!
+        bconst = price_ref / 1e3 * (gdp0 / demand_ref) ** (rho - 1)
+        self.data['bconst'] = bconst
+        return self.data['bconst']
+
+    @lru_cache()
+    def _aconst(self):
+        bconst = self._bconst()
+        demand_ref = self.data['demand_ref']
+        rho = self._rho()
+        gdp0 = self._gdp0()
+        k0 = self._k0()
+        kpvs = self.data['kpvs']
+        # TODO: why name this partmp??
+        partmp = (bconst * demand_ref ** rho).groupby(level='node').sum()
+        # TODO: automatically get the units here!!
+        aconst = ((gdp0 / 1e3) ** rho - partmp) / (k0 / 1e3) ** (rho * kpvs)
+        self.data['aconst'] = aconst
+        return self.data['aconst']
+
 
 def init(s):
-    for args in MACRO_INIT['sets']:
-        s.init_set(*args)
-    for args in MACRO_INIT['pars']:
-        s.init_par(*args)
-    # for args in MACRO_INIT['vars']:
-    #     print(args)
-    #     s.init_var(*args)
-    # for args in MACRO_INIT['equs']:
-    #     s.init_equ(*args)
+    for key, values in MACRO_INIT['sets'].items():
+        s.init_set(key, values)
```

Comment:
@gidden, if a scenario already has MACRO-related sets, parameters, variables and equations; there will be an error when initializing these items here. 
1) To resolve this issue, we can add a check here for each item, e.g. for sets:
`if not s.has_set(key): s.init_set(key, values)`
2) On the other hand, if a scenario has a MACRO-related data, we may want to remove that data from the scenario before adding new data (in case the existing data is not correct or relevant anymore). So, we may want to remove these items in the first place and initialize them again.

Reply:
Hey @behnam2015, good points here. My initial reaction is that if a scenario already has MACRO-related data, then it *should* be an error to try to "re-add" it. Perhaps we can add an additional function to `remove()` MACRO related parameters as well. 

This is a bit tricky, because I observed that even in the westeros model, some things already existed (e.g., the set `mapping_macro_sector`), which I believe could be added automatically on the java side of things. Can @danielhuppmann chime in with his thoughts here?

Reply:
I did a quick test to remove things and then re-add, but got the following error:

```
self = <message_ix.core.Scenario object at 0x7efd66bfd0f0>, name = 'historical_gdp', idx_sets = ['node', 'year'], idx_names = None

    def init_par(self, name, idx_sets, idx_names=None):
        """Initialize a new parameter.
    
        Parameters
        ----------
        name : str
            Name of the parameter.
        idx_sets : list of str
            Names of sets that index this parameter.
        idx_names : list of str, optional
            Names of the dimensions indexed by `idx_sets`.
        """
>       self._jobj.initializePar(name, *make_dims(idx_sets, idx_names))
E       jpype._jclass.at.ac.iiasa.ixmp.exceptions.IxException: Parameter 'historical_gdp' can only be initialized with MESSAGE standard dimensions!
```

Reply:
@behnam2015, I have made a new branch on my remote called `macro-behnam-test`, which includes a new try/catch for `init_set`. Let's see what @danielhuppmann says before we do anything on this PR branch. But in the meantime, you can use `macro-behnam-test` to start testing the global model (it includes the `ignore_node` and `ignore_sector` feature as well).

Reply:
Thanks @gidden. By removing data, I meant only removing the content of a parameter and not the parameter itself. Something like:
key = 'historical_gdp'
```
if s.has_par(key):
    s.remove_par(key, s.par(key))
else:
    s.init_par(key, idx_sets)
```